File: Data Structures & Algorithms/trapping-rain-water/submission-6.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int[] rightMax = new int[n];
        for(int i=n-2; i>=0; i--) {
            rightMax[i] = Math.max(rightMax[i+1], height[i+1]);
        }
        int ans = 0;
        int left = 0;
        for(int i=0; i<n; i++) {
            int limit = Math.min(left, rightMax[i]);
            int tmp = Math.max(0, limit-height[i]);
            ans += tmp;
            left = Math.max(left, height[i]);
        }
        return ans;
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-6.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        currmax, currmin = nums[0], nums[0]
        ans = nums[0]
        for i in range(1,len(nums)):
            tmp = currmax
            currmax = max(nums[i], currmin*nums[i], currmax*nums[i])
            currmin = min(nums[i], currmin*nums[i], tmp*nums[i])
            ans = max(ans, currmax)
        return ans